File: mathsys/dev/python/level5.py
#^
#^  HEAD
#^

#> HEAD -> MODULES
from abc import ABC
from dataclasses import dataclass
from typing import cast

#> HEAD -> DATA
from .nonterminal import NonTerminal
from .tokenizer import IDENTIFIER, NUMBER, RATIONAL, TYPE
from .ir import Opcode, node, Binary, Pointer, String, Option, Vec, BigUint, Group
from .latex import VARIABLES, CONVERSION


#^
#^  5ºLEVEL
#^

#> 5ºLEVEL -> NAMESPACE
class Level5(NonTerminal, ABC): pass

#> 5ºLEVEL -> INFINITE
@dataclass(init = False, unsafe_hash = True)
class Infinite(Level5):
    code = Opcode(0x0C).binary()
    def __init__(self, items: list) -> None: self.freeze()
    def latex(self, types: dict[str, str]) -> str: return r"\infty "
    def ir(self, binary: list[Binary], nodes: list[Binary]) -> Pointer:
        return node(self.code, binary, nodes)

#> 5ºLEVEL -> VARIABLE
@dataclass(init = False, unsafe_hash = True)
class Variable(Level5):
    code = Opcode(0x0D).binary()
    representation: str
    def __init__(self, items: list[IDENTIFIER]) -> None: self.representation = items[0].value; self.freeze()
    def latex(self, types: dict[str, str]) -> str:
        curated = self.representation
        for source, replace in VARIABLES.items(): curated = curated.replace(source, replace)
        return CONVERSION[types.get(self.representation, "@Undefined")](curated)
    def ir(self, binary: list[Binary], nodes: list[Binary]) -> Pointer:
        representation = String(self.representation)
        return node(self.code + representation, binary, nodes)

#> 5ºLEVEL -> NEST
@dataclass(init = False, unsafe_hash = True)
class Nest(Level5):
    code = Opcode(0x0E).binary()
    value: Level2 | None
    def __init__(self, items: list[Level2]) -> None: 
        self.value = items[0] if len(items) == 1 else None
        self.freeze()
    def latex(self, types: dict[str, str]) -> str:
        inside = self.value.latex(types) if self.value is not None else ""
        return fr"\left( {inside}\right) "
    def ir(self, binary: list[Binary], nodes: list[Binary]) -> Pointer:
        value = Option(self.value.ir(binary, nodes) if self.value is not None else None)
        return node(self.code + value, binary, nodes)

#> 5ºLEVEL -> TENSOR
@dataclass(init = False, unsafe_hash = True)
class Tensor(Level5):
    code = Opcode(0x0F).binary()
    values: tuple[Level2, ...]
    def __init__(self, items: list[Level2]) -> None: self.values = tuple(items); self.freeze()
    def latex(self, types: dict[str, str]) -> str:
        inside = r"\; " if len(self.values) == 0 else r"\\ ".join(value.latex(types) for value in self.values)
        return fr"\begin{{bmatrix}}{inside}\end{{bmatrix}}"
    def ir(self, binary: list[Binary], nodes: list[Binary]) -> Pointer:
        values = Vec([value.ir(binary, nodes) for value in self.values])
        return node(self.code + values, binary, nodes)

#> 5ºLEVEL -> WHOLE
@dataclass(init = False, unsafe_hash = True)
class Whole(Level5):
    code = Opcode(0x10).binary()
    value: int
    def __init__(self, items: list[NUMBER]) -> None: self.value = int(items[0].value); self.freeze()
    def latex(self, types: dict[str, str]) -> str: return str(self.value)
    def ir(self, binary: list[Binary], nodes: list[Binary]) -> Pointer:
        value = BigUint(self.value)
        return node(self.code + value, binary, nodes)

#> 5ºLEVEL -> ABSOLUTE
@dataclass(init = False, unsafe_hash = True)
class Absolute(Level5):
    code = Opcode(0x11).binary()
    value: Level2
    def __init__(self, items: list[Level2]) -> None: self.value = items[0]; self.freeze()
    def latex(self, types: dict[str, str]) -> str: return fr"\left| {self.value.latex(types)}\right| "
    def ir(self, binary: list[Binary], nodes: list[Binary]) -> Pointer:
        value = self.value.ir(binary, nodes)
        return node(self.code + value, binary, nodes)

#> 5ºLEVEL -> UNDEFINED
@dataclass(init = False, unsafe_hash = True)
class Undefined(Level5):
    code = Opcode(0x12).binary()
    def __init__(self, items: list) -> None: self.freeze()
    def latex(self, types: dict[str, str]) -> str: return r"\left. ?\right. "
    def ir(self, binary: list[Binary], nodes: list[Binary]) -> Pointer: return node(self.code, binary, nodes)

#> 5ºLEVEL -> RATIONAL
@dataclass(init = False, unsafe_hash = True)
class Rational(Level5):
    code = Opcode(0x13).binary()
    whole: int
    decimal: int
    def __init__(self, items: list[RATIONAL]) -> None:
        self.whole = int(items[0].value.split(".", 1)[0])
        self.decimal = int(items[0].value.split(".", 1)[1].rstrip("0")[::-1])
        self.freeze()
    def latex(self, types: dict[str, str]) -> str: return f"{self.whole}.{str(self.decimal)[::-1]}"
    def ir(self, binary: list[Binary], nodes: list[Binary]) -> Pointer:
        whole = BigUint(self.whole)
        decimal = BigUint(self.decimal)
        return node(self.code + whole + decimal, binary, nodes)

#> 5ºLEVEL -> CASTS
@dataclass(init = False, unsafe_hash = True)
class Casts(Level5):
    code = Opcode(0x14).binary()
    element: Level5
    to: str
    def __init__(self, items: list[Level5 | TYPE]) -> None:
        self.element = cast(Level5, items[0])
        self.to = cast(TYPE, items[1]).value
        self.freeze()
    def latex(self, types: dict[str, str]) -> str: return self.element.latex(types)
    def ir(self, binary: list[Binary], nodes: list[Binary]) -> Pointer:
        element = self.element.ir(binary, nodes)
        to = Group(self.to)
        return node(self.code + element + to, binary, nodes)

#> 5ºLEVEL -> TYPES
from .level2 import Level2

File: mathsys/dev/typescript/level5.ts
//^
//^ HEAD
//^

//> HEAD -> DATA
import {NonTerminal} from "./nonterminal";
import {IDENTIFIER, NUMBER, RATIONAL, TYPE} from "./tokenizer";
import {Opcode, node, Binary, Pointer, String as Str, Option, Vec, BigUint, Group} from "./ir";
import {VARIABLES, CONVERSION} from "./latex";


//^
//^ 5ºLEVEL
//^

//> 5ºLEVEL -> NAMESPACE
export abstract class Level5 extends NonTerminal {}

//> 5ºLEVEL -> INFINITE
export class Infinite extends Level5 {
    private code = new Opcode(0x0C).binary();
    constructor(items: []) {super(); this.freeze(); return this}
    latex(types: Map<string, string>): string {return String.raw`\infty `}
    ir(binary: Binary[], nodes: Binary[]): Pointer {return node(this.code, binary, nodes)}
}

//> 5ºLEVEL -> VARIABLE
export class Variable extends Level5 {
    private code = new Opcode(0x0D).binary();
    readonly representation: string;
    constructor(items: IDENTIFIER[]) {
        super();
        this.representation = items[0].value;
        this.freeze();
        return this;
    }
    latex(types: Map<string, string>): string {
        let curated = this.representation;
        for (const [source, replace] of VARIABLES) {
            curated = curated.replaceAll(source, replace);
        }
        return (CONVERSION.get(types.get(this.representation) ?? "@Undefined") as (name: string) => string)(curated)
    }
    ir(binary: Binary[], nodes: Binary[]): Pointer {
        const representation = new Str(this.representation);
        return node(this.code.add(representation), binary, nodes);
    }
}

//> 5ºLEVEL -> NEST
export class Nest extends Level5 {
    private code = new Opcode(0x0E).binary();
    readonly value: Level2 | null;
    constructor(items: Level2[]) {
        super();
        this.value = items.length === 1 ? items[0] : null;
        this.freeze();
        return this;
    }
    latex(types: Map<string, string>): string {
        const inside = this.value !== null ? this.value.latex(types) : "";
        return String.raw`\left( ${inside}\right) `;
    }
    ir(binary: Binary[], nodes: Binary[]): Pointer {
        const value = new Option(this.value !== null ? this.value.ir(binary, nodes) : null);
        return node(this.code.add(value), binary, nodes);
    }
}

//> 5ºLEVEL -> TENSOR
export class Tensor extends Level5 {
    private code = new Opcode(0x0F).binary();
    readonly values: readonly Level2[];
    constructor(items: Level2[]) {
        super();
        this.values = items;
        this.freeze();
        return this;
    }
    latex(types: Map<string, string>): string {
        const inside = this.values.length === 0 ? String.raw`\; ` : this.values.map(value => value.latex(types)).join(String.raw`\\ `);
        return String.raw`\begin{bmatrix}${inside}\end{bmatrix}`;
    }
    ir(binary: Binary[], nodes: Binary[]): Pointer {
        const values = new Vec(this.values.map(value => value.ir(binary, nodes)));
        return node(this.code.add(values), binary, nodes);
    }
}

//> 5ºLEVEL -> WHOLE
export class Whole extends Level5 {
    private code = new Opcode(0x10).binary();
    readonly value: bigint;
    constructor(items: NUMBER[]) {
        super();
        this.value = BigInt(items[0].value);
        this.freeze();
        return this;
    }
    latex(types: Map<string, string>): string {return this.value.toString()}
    ir(binary: Binary[], nodes: Binary[]): Pointer {
        const value = new BigUint(this.value);
        return node(this.code.add(value), binary, nodes);
    }
}

//> 5ºLEVEL -> ABSOLUTE
export class Absolute extends Level5 {
    private code = new Opcode(0x11).binary();
    readonly value: Level2;
    constructor(items: Level2[]) {
        super();
        this.value = items[0];
        this.freeze();
        return this;
    }
    latex(types: Map<string, string>): string {return String.raw`\left| ${this.value.latex(types)}\right| `}
    ir(binary: Binary[], nodes: Binary[]): Pointer {
        const value = this.value.ir(binary, nodes);
        return node(this.code.add(value), binary, nodes);
    }
}

//> 5ºLEVEL -> UNDEFINED
export class Undefined extends Level5 {
    private code = new Opcode(0x12).binary();
    constructor(items: []) {
        super();
        this.freeze();
        return this;
    }
    latex(types: Map<string, string>): string {return String.raw`\left. ?\right. `}
    ir(binary: Binary[], nodes: Binary[]): Pointer {return node(this.code, binary, nodes)}
}

//> 5ºLEVEL -> RATIONAL
export class Rational extends Level5 {
    private code = new Opcode(0x13).binary();
    readonly whole: bigint;
    readonly decimal: bigint;
    constructor(items: RATIONAL[]) {
        super();
        this.whole = BigInt(items[0].value.split(".", 1)[0]);
        this.decimal = BigInt(items[0].value.split(".", 1)[1].replace(/0*$/, "").split("").reverse().join(""));
        this.freeze();
        return this;
    }
    latex(types: Map<string, string>): string {return `${this.whole}.${this.decimal.toString().split("").reverse().join("")}`}
    ir(binary: Binary[], nodes: Binary[]): Pointer {
        const whole = new BigUint(this.whole);
        const decimal = new BigUint(this.decimal);
        return node(this.code.add(whole, decimal), binary, nodes);
    }
}

//> 5ºLEVEL -> CASTS
export class Casts extends Level5 {
    private code = new Opcode(0x014).binary();
    readonly element: Level5;
    readonly to: string;
    constructor(items: (Level5 | TYPE)[]) {
        super();
        this.element = items[0] as Level5;
        this.to = (items[1] as TYPE).value;
        this.freeze();
        return this;
    }
    latex(types: Map<string, string>): string {return this.element.latex(types)}
    ir(binary: Binary[], nodes: Binary[]): Pointer {
        const element = this.element.ir(binary, nodes);
        const to = new Group(this.to);
        return node(this.code.add(element, to), binary, nodes);
    }
}

//> 5ºLEVEL -> TYPES
import {Level2} from "./level2";